### DISCUSSION TITLE
Turtle api question.

### DISCUSSION THREAD

**Discussion by SirEndii** [cc-tweaked/CC-Tweaked#796] [Q&A] [answered]
Hello, I have problems with some Turtles. When I create one turtle and try to get the position of the turtle via a self-made peripheral, the position is correct. But when I place a second Turtle and try to get the position via a self-made peripheral, it returns the position of the first Turtle. 

More context: https://github.com/Seniorendi/AdvancedPeripherals/issues/54

I hope anyone here can understand what I mean. It's like, that every turtle use the same instance of my turtle.

**Comment by SquidDev** [comment, marked as answer]
I can't say for sure, but this block looks pretty concerning:

https://github.com/Seniorendi/AdvancedPeripherals/blob/49fe99e025de770f4a02ab37d502ee007d3b973f/src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/base/BaseTurtle.java#L66-L74

That'll only get triggered once, so if the turtle ever moves you'll have a stale tile entity. Ideally this code wouldn't exist at all and you'd just use `ITurtleAccess`, though I understand why you're doing this.

That said, not sure why turtle instances are merging into one - nothing in the code looks suspect there.

**Reply by SirEndii** [reply]
I changed the system a little. I now use a var `ITurtleAcces` in my BasePeripheral class. I can set the turtle with a `setTurtle` method.
Because `tileEntity` and `entity` is null, my turtle uses the turtle var to get the position.
Enabled the debug mode and the turtle does not tick? The update method does not fire because the init boolean is true. The init boolean is for every turtle the same. 

Debug1 and Debug2 does not show up. But I debug another message when I use the function `isSlimeChunk` of the `EnvironmentDetectorPeripheral` class. `AdvancedPeripherals.debug("DEBUG3 " + getPos());` When I try to use isSlimeChunk with a pocket computer, it works. But When I use it with a turtle, it returns a weird block pos `BlockPos{x=-22, y=98, z=16}`. At this position is nothing, just air over a desert. 

New update method:
```java
    @Override
    public void update(@NotNull ITurtleAccess turtle, @NotNull TurtleSide side) {
        if (!init) {
            if (!turtle.getWorld().isRemote) {
                AdvancedPeripherals.debug("DEBUG1 " + turtle.getPosition());
                AdvancedPeripherals.debug("DEBUG2 " + turtle);
                peripheral.setTurtle(turtle);
                init = true;
            }
        }
        tick++;
        if (tick > 10) {
            TileEntityList tileEntityList = TileEntityList.get(turtle.getWorld()); //Sync the position with the tile entity list.
            tileEntityList.setTileEntity(turtle.getWorld(), turtle.getPosition(), true); //Add the turtle to the List for event use
            tick = 0;
        }
    }
```

**Reply by SquidDev** [reply]
Oh, I'm very silly. Sorry, should have spotted this earlier.

`ITurtleAccess` is a singleton, and so your `BaseTurtle.peripheral` variable will be shared across all turtles. Instead of caching it, you should be overriding `createPeripheral(ITurtleAccess, TurtleSide)` to return a `new EnvironmentDetectorPeripheral(turtle)` (or similar).

If you need to access the peripheral instance in the `update` method, you can do something like we do for modems (agreed it's a little ugly - I should probably refactor the interface next Minecraft version):

https://github.com/SquidDev-CC/CC-Tweaked/blob/58ea7a275e70e69f29641adc3d8a3ce59b61b068/src/main/java/dan200/computercraft/shared/turtle/upgrades/TurtleModem.java#L133-L134

**Reply by SirEndii** [reply]
Thank you! Everything works now fine
